File: GildedRose/src/com/gildedrose/tests/ConjuredItemTest.java
package com.gildedrose.tests;

import com.gildedrose.items.*;
import com.gildedrose.GildedRose;

import org.junit.Test;
import static org.junit.Assert.*;


public class ConjuredItemTest {

    @Test
    public void updateQuality_SellinGreterThanZero_QualityDecreaseByTwo() {
    	//Arrange
        Item[] items = new Item[] { new Item("Conjured Mana Cake", 10, 20) };
        GildedRose app = new GildedRose(items);
        
        //Act
        app.updateQuality();
        
        //Assert
        assertEquals("Conjured Mana Cake", items[0].name);
        assertEquals(9, items[0].sellIn);
        assertEquals(18, items[0].quality);
    }

    @Test
    public void updateQuality_SellinEqualsZero_QualityDecreaseByFour() {
    	//Arrange
        Item[] items = new Item[] { new Item("Conjured Mana Cake", 0, 20) };
        GildedRose app = new GildedRose(items);
        
        //Act
        app.updateQuality();
        
        //Assert
        assertEquals("Conjured Mana Cake", items[0].name);
        assertEquals(-1, items[0].sellIn);
        assertEquals(16, items[0].quality);
    }

    @Test
    public void updateQuality_SellinLessThanZero_QualityDecreaseByFour() {
    	//Arrange
        Item[] items = new Item[] { new Item("Conjured Mana Cake", -1, 20) };
        GildedRose app = new GildedRose(items);
        
        //Act
        app.updateQuality();
        
        //Assert
        assertEquals("Conjured Mana Cake", items[0].name);
        assertEquals(-2, items[0].sellIn);
        assertEquals(16, items[0].quality);
    }

    @Test
    public void updateQuality_QualityEqualsZero_QualityUnchanged() {
    	//Arrange
        Item[] items = new Item[] { new Item("Conjured Mana Cake", 10, 0) };
        GildedRose app = new GildedRose(items);
        
        //Act
        app.updateQuality();
        
        //Assert
        assertEquals("Conjured Mana Cake", items[0].name);
        assertEquals(9, items[0].sellIn);
        assertEquals(0, items[0].quality);
    }

    @Test
    public void updateQuality_QualityEqualsFifty_QualityDecreaseByTwo() {
    	//Arrange
        Item[] items = new Item[] { new Item("Conjured Mana Cake", 10, 50) };
        GildedRose app = new GildedRose(items);
        
        //Act
        app.updateQuality();
        
        //Assert
        assertEquals("Conjured Mana Cake", items[0].name);
        assertEquals(9, items[0].sellIn);
        assertEquals(48, items[0].quality);
    }

    @Test
    public void updateQuality_QualityEqualsThree_QualityToZero() {
    	//Arrange
        Item[] items = new Item[] { new Item("Conjured Mana Cake", 0, 3) };
        GildedRose app = new GildedRose(items);
        
        //Act
        app.updateQuality();
        
        //Assert
        assertEquals("Conjured Mana Cake", items[0].name);
        assertEquals(-1, items[0].sellIn);
        assertEquals(0, items[0].quality);
    }

}

File: GildedRose.Net/Tests/ConjuredItemTest.cs
using Xunit;

using GildedRose.Items;

namespace GildedRose.Tests
{
    public class ConjuredItemTest
    {
        
        [Fact] 
        public void updateQuality_SellinGreterThanZero_QualityDecreaseByTwo() {
            //Arrange
            Item[] items = new Item[] { new Item{Name = "Conjured Mana Cake", SellIn=10, Quality = 20} };
            GildedRose app = new GildedRose(items);
            
            //Act
            app.UpdateQuality();
            
            //Assert
            Assert.Equal("Conjured Mana Cake", items[0].Name);
            Assert.Equal(9, items[0].SellIn);
            Assert.Equal(18, items[0].Quality);
        }
    
        [Fact]
        public void updateQuality_SellinEqualsZero_QualityDecreaseByFour() {
            //Arrange
            Item[] items = new Item[] { new Item{Name = "Conjured Mana Cake", SellIn=0, Quality = 20} };
            GildedRose app = new GildedRose(items);
            
            //Act
            app.UpdateQuality();
            
            //Assert
            Assert.Equal("Conjured Mana Cake", items[0].Name);
            Assert.Equal(-1, items[0].SellIn);
            Assert.Equal(16, items[0].Quality);
        }
    
        [Fact]
        public void updateQuality_SellinLessThanZero_QualityDecreaseByFour() {
            //Arrange
            Item[] items = new Item[] { new Item{Name = "Conjured Mana Cake", SellIn=-1, Quality = 20} };
            GildedRose app = new GildedRose(items);
            
            //Act
            app.UpdateQuality();
            
            //Assert
            Assert.Equal("Conjured Mana Cake", items[0].Name);
            Assert.Equal(-2, items[0].SellIn);
            Assert.Equal(16, items[0].Quality);
        }
    
        [Fact]
        public void updateQuality_QualityEqualsZero_QualityUnchanged() {
            //Arrange
            Item[] items = new Item[] { new Item{Name = "Conjured Mana Cake", SellIn=10, Quality = 0} };
            GildedRose app = new GildedRose(items);
            
            //Act
            app.UpdateQuality();
            
            //Assert
            Assert.Equal("Conjured Mana Cake", items[0].Name);
            Assert.Equal(9, items[0].SellIn);
            Assert.Equal(0, items[0].Quality);
        }
    
        [Fact]
        public void updateQuality_QualityEqualsFifty_QualityDecreaseByTwo() {
            //Arrange
            Item[] items = new Item[] { new Item{Name = "Conjured Mana Cake", SellIn=10, Quality = 50} };
            GildedRose app = new GildedRose(items);
            
            //Act
            app.UpdateQuality();
            
            //Assert
            Assert.Equal("Conjured Mana Cake", items[0].Name);
            Assert.Equal(9, items[0].SellIn);
            Assert.Equal(48, items[0].Quality);
        }
    
        [Fact]
        public void updateQuality_QualityEqualsThree_QualityToZero() {
            //Arrange
            Item[] items = new Item[] { new Item{Name = "Conjured Mana Cake", SellIn=0, Quality = 3} };
            GildedRose app = new GildedRose(items);
            
            //Act
            app.UpdateQuality();
            
            //Assert
            Assert.Equal("Conjured Mana Cake", items[0].Name);
            Assert.Equal(-1, items[0].SellIn);
            Assert.Equal(0, items[0].Quality);
        }

    }
}
